core/src/test/java/org/apache/calcite/tools/PlannerTest.java

```diff
@@ -1030,10 +1030,12 @@ private void checkJoinNWay(int n) throws Exception {
         + "EnumerableProject(empid=[$2], deptno=[$3], name=[$4], salary=[$5], commission=[$6], deptno0=[$7], name0=[$8], employees=[$9], location=[ROW($10, $11)], empid0=[$0], name1=[$1])\n"
         + "  EnumerableHashJoin(condition=[=($0, $2)], joinType=[left])\n"
         + "    EnumerableTableScan(table=[[hr, dependents]])\n"
-        + "    EnumerableHashJoin(condition=[=($1, $5)], joinType=[inner])\n"
-        + "      EnumerableTableScan(table=[[hr, emps]])\n"
-        + "      EnumerableProject(deptno=[$0], name=[$1], employees=[$2], x=[$3.x], y=[$3.y])\n"
-        + "        EnumerableTableScan(table=[[hr, depts]])";
+        + "    EnumerableMergeJoin(condition=[=($1, $5)], joinType=[inner])\n"
```

Comment:
Why hash join is switched to merge join? It looks a bit weird because inputs are not sorted and we have to sort them explicitly which is a heavy task. Can it be a sign of some bug?

Reply:
It is chosen by cost model, the new plan is cheaper:
![image](https://user-images.githubusercontent.com/15352793/81355556-78889d00-9094-11ea-8cb0-87d473c719fb.png)
Calcite's default cost model is very arbitrary.

Reply:
It seems to be you are right. The problem is in the cost model: `VolcanoCost` is compared only by `rows` number. And not clear (or outdated) cost estimation in `EnumerableHashJoin#computeSelfCost`.